code/gateway/deregister.q

```diff
@@ -0,0 +1,43 @@
+
+\d .gw
+
+//overwrite this method to not upsert null handles
+addserversfromconnectiontable:{
+ {.gw.addserverattr'[x`w;x`proctype;x`attributes]}[select w,proctype,attributes from .servers.SERVERS where ((proctype in x) or x~`ALL),not w in ((0;0Ni),exec handle from .gw.servers where not null handle)];}
```

Comment:
why does this need overwritten? are there cases were a server is "active" but has a null handle?

Reply:
Hi Jonathan,
    Actually, its for the opposite case. There will be cases where a server has a valid handle but is not active. The standard method will add a new entry with a new serverid, same handle, active flag set to true for all non-active servers (this is called each time a new server connection is established)
    The changes in this PR want to prevent servers with the active flag set to false from being used. So, I instead only want to add new records to .gw.servers where the handle is invalid (otherwise we would have multiple entries in ```.gw.servers``` for the same handle)
    This is actually why, in the TorQ changes I ignored the active clause below:

```
   serverid:$[@[value;`.finspace.enabled;0b];
    first exec serverid from .gw.servers where handle=.z.w;
    first exec serverid from .gw.servers where active, handle=.z.w];
```

   I believe this check was originally there because of cases where there were multiple entries in ```.gw.servers``` using the same handle, so one would want to select the entry where the active flag was set to true. By overwriting this method, there will only be one entry in ```.gw.servers``` per handle, so there is no need to check the active clause.
    For example:
        query gets sent to serverid 2 using handle 1i. serverid 2's active flag is set to false after the query is sent. When the response returns, we check ``` .gw.servers ``` for where active, handle=1i. This will map the serverid to null. the gateway never thinks the query is completed because ```.gw.results``` is not updated appropriately. If we ignore the active clause, we can map the serverid to 2 (which is what we want)
    Example 2 (This is with standard functionality)
        serverids 4 and 6 are using handle 1i, but serverid 4 has the active flag set to false while serverid 6 was active flag set to true. Query comes in and gw updates ```.gw.results``` with information about the serverid, servertype, client handle, message to send client, whether the query completed, etc. Query gets returned, so we check ```.gw.servers``` where active, handle = 1i - this will map to serverid 6 and successfully ignore serverid 4 (which is what we want)